File: Python/Graphs/Prims/impl.py
from typing import List
from collections import defaultdict
import heapq


"""
    The following has a similar time and space complexity to Dijkstra's at O(E) space and O(E * log(V)) time.
"""

class Solution:
    def minimumSpanningTree(self, n: int, edges: List[List[int]]) -> int:
        
        # create the adj_m O(E) time and space

        adj_m = defaultdict(list)

        for v1, v2, w in edges:
            adj_m[v1].append((v2, w))
            adj_m[v2].append((v1, w))
        
        heap = [[0, 0]] # arbitrarly choosing an edge. Can be edges[0][0]
        visited = set()
        res = 0

        while heap and len(visited) < n: #if making a list, the condidion is while the length of the list is not the same as edges
            w, v = heapq.heappop(heap)
            if v in visited:
                continue
            visited.add(v)
            res += w

            for v2, w2 in adj_m[v]:
                if v2 not in visited:
                    heapq.heappush(heap, (w2, v2))

        
        return res if len(visited) == n else -1

File: c++/graphs/Prims/impl.cpp
#include <algorithm>
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

/**
 * Prims algorithm is used to find the minimum spanning tree. The time complexity of this is O(E * log(v)) and the space is O(E).
 */

class Solution
{
public:
    int minimumSpanningTree(vector<vector<int>> &edges, int n)
    {

        // create the adjcency list
        unordered_map<int, vector<pair<int, int>>> adj_l;

        for (const auto &edge : edges)
        {
            adj_l[edge[0]].push_back({edge[1], edge[2]});
            adj_l[edge[1]].push_back({edge[0], edge[2]});
        }

        unordered_set<int> visited;
        vector<pair<int, int>> heap;
        heap.push_back({0, 0});
        int sum = 0;

        while (visited.size() < n && heap.size() > 0)
        {
            int w = heap.front().first;
            int src = heap.front().second;

            pop_heap(heap.begin(), heap.end());
            heap.pop_back();

            if (visited.count(src) == 0)
            {
                sum -= w;
                visited.insert(src);

                for (const auto &c : adj_l[src])
                {
                    int dst = c.first;
                    int w2 = c.second;
                    if (visited.count(dst) == 0)
                    {
                        heap.push_back({-w2, dst});
                        push_heap(heap.begin(), heap.end());
                    }
                }
            }
        }

        return (visited.size() == n) ? sum : -1;
    }
};
